alternator/scoped_item_list.hh

```diff
@@ -0,0 +1,181 @@
+ */
+
+/**
+ * scoped_item_list<T> holds a list of items of type T, which are
+ * automatically removed from the list when they go out of scope.
+ * This is useful, for example, to hold a list of ongoing requests in a server
+ * (on one particular shard), so that they can be listed by an API that wants
+ * to list such requests (e.g., the "system.clients" virtual table).
+ *
+ * When adding an item to the list with the emplace() method, a handle is
+ * returned which can be used to continue accessing the item. When this handle
+ * goes out of scope, the item is automatically removed from the list.
+ * 
+ * Importantly, scoped_item_list<T> also provides for_each_gently() - a
+ * stall-safe way to iterate over the list. for_each_gently() run a function
+ * on each item in the list, but can preempt between items when our time quota
+ * is finished, to avoid stalls.
+ * 
+ * For example, code that tracks ongoing requests in a server might look
+ * like this:
+ *
+ * scoped_item_list<request_data> ongoing_requests;
+ *
+ * In the handler of each incoming request:
+ *     auto req_handle = ongoing_requests.emplace(...);
+ *     // the handler continues to run and perform its work.
+ *     // If it wishes, it can use and modify req_handle.
+ *     ...
+ *     // when the handler is done, req_handle goes out of scope and the
+ *     // request's entry is automatically removed from ongoing_requests.
+ *
+ * In the API that lists the ongoing requests:
+ *     co_await ongoing_requests.for_each_gently([] (request_data& req) {
+ *         // Do something with req, e.g., format it and add it to a
+ *         // a chunked_vector.
+ *         ...
+ *      });
+ *
+ * NOTE:
+ * scoped_item_list<T> currently assumes that when it is destroyed, it is
+ * already empty, i.e., all handles returned by it had already been destroyed.
+ * If this is not the case, on_fatal_internal_error() is called.
+ * This assumption holds when used in the Alternator server, where the server
+ * and the scoped_item_list<T> it contains, are only destroyed after
+ * waiting for the _pending_requests gate to close, and it cannot close
+ * before all calls to handle_api_request() are done. And since
+ * handle_api_request() is the only place that holds scoped_item_list<T>
+ * handles, when all these calls finished no handles can be held.
+ * 
+ * This class is inspired by the generic_server::connection class,
+ * which uses similar code to hold ongoing requests in a list, but
+ * is too tightly integrated into generic_server, which Alternator
+ * cannot use because it uses Seastar's HTTP server for managing the
+ * connections.
+ */
+
+#pragma once
+
+
+#include <list>
+
+#include <boost/intrusive/list.hpp>
+
+#include <seastar/core/future.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/util/noncopyable_function.hh>
+
+#include "seastarx.hh"
+#include "utils/on_internal_error.hh"
+
+template <typename T>
+class scoped_item_list {
```

Comment:
There's nothing alternator specific here, it should be in utils/ and come with its own test.